Authentication fails when HTTPS is enabled

I have followed the explanations given in the [documentation](http://doc.activeeon.com/latest/admin/ProActiveAdminGuide.html#_enable_https) to configure the Scheduler so that Jetty serves requests using HTTPS.

In summary, I have edited `config/web/settings.ini` in order to set `web.https` to `true`. 

Once done, the Scheduler and all services start properly. However, it is not possible to authenticate to RM or Scheduler Web portals (no error message in Scheduler logs).

![capture d ecran de 2016-02-28 16-54-56](https://cloud.githubusercontent.com/assets/128898/13380215/f06e91c0-de3d-11e5-992c-f21d08575e4a.png)

I remember that unpacked WAR files associated to RM and Scheduler Web apps contain configuration files. Consequently, I have tried to edit them:

  - `dist/war/rm/rm.conf ` to set `rm.rest.url=https://localhost:8080/rest`
  - `dist/war/scheduler/scheduler.conf` to set `sched.rest.url=https://localhost:8080/rest`

After applying these configurations, the Scheduler starts but authentication is still not working (endless connection):

![capture d ecran de 2016-02-28 16-57-50](https://cloud.githubusercontent.com/assets/128898/13380268/ef1dbf5c-de3e-11e5-9d3c-20977d3a43a3.png)

When I look at the logs on Scheduler side, I get the following error message:

```
[2016-02-28 16:57:41,718 INFO  .p.s.a.SchedulerAuthentication] admin is trying to connect
[2016-02-28 16:57:41,718 INFO  .p.s.a.SchedulerAuthentication] User admin logged successfully
[2016-02-28 16:57:41,718 INFO  .p.s.a.SchedulerAuthentication] user : admin
[2016-02-28 16:57:41,719 INFO     o.o.p.s.c.SchedulingService] admin successfully connected !
[2016-02-28 16:57:41,727 INFO     o.o.p.s.c.SchedulingService] USER space for user admin is at [file:/home/lpellegr/Bureau/HODAC%20PoC/activeeon-pws-enterprise-7.3.1-windows-x64/data/defaultuser/admin, pappnp://leonard:64738/UserSpaces?proactive_vfs_provider_path=/admin]
[2016-02-28 16:57:41,836 WARN                      /scheduler] Exception while dispatching incoming RPC call
com.google.gwt.user.server.rpc.UnexpectedException: Service method 'public abstract java.lang.String org.ow2.proactive_grid_cloud_portal.scheduler.client.SchedulerService.getVersion() throws org.ow2.proactive_grid_cloud_portal.common.shared.RestServerException,org.ow2.proactive_grid_cloud_portal.common.shared.ServiceException' threw an unexpected exception: javax.ws.rs.ProcessingException: RESTEASY004655: Unable to invoke request
	at com.google.gwt.user.server.rpc.RPC.encodeResponseForFailure(RPC.java:415)
	at com.google.gwt.user.server.rpc.RPC.invokeAndEncodeResponse(RPC.java:605)
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processCall(RemoteServiceServlet.java:333)
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processCall(RemoteServiceServlet.java:303)
	at com.google.gwt.user.server.rpc.RemoteServiceServlet.processPost(RemoteServiceServlet.java:373)
	at com.google.gwt.user.server.rpc.AbstractRemoteServiceServlet.doPost(AbstractRemoteServiceServlet.java:62)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:587)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:577)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.ws.rs.ProcessingException: RESTEASY004655: Unable to invoke request
	at org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine.invoke(ApacheHttpClient4Engine.java:287)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:436)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:102)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:64)
	at com.sun.proxy.$Proxy253.getVersion(Unknown Source)
	at org.ow2.proactive_grid_cloud_portal.scheduler.server.SchedulerServiceImpl$26.apply(SchedulerServiceImpl.java:918)
	at org.ow2.proactive_grid_cloud_portal.scheduler.server.SchedulerServiceImpl$26.apply(SchedulerServiceImpl.java:915)
	at org.ow2.proactive_grid_cloud_portal.scheduler.server.SchedulerServiceImpl.executeFunctionReturnStreamAsString(SchedulerServiceImpl.java:1113)
	at org.ow2.proactive_grid_cloud_portal.scheduler.server.SchedulerServiceImpl.getVersion(SchedulerServiceImpl.java:915)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.gwt.user.server.rpc.RPC.invokeAndEncodeResponse(RPC.java:587)
	... 25 more
Caused by: javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1949)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:302)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:296)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1509)
	at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:216)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:979)
	at sun.security.ssl.Handshaker.process_record(Handshaker.java:914)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1062)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
	at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:535)
	at org.apache.http.conn.ssl.SSLSocketFactory.connectSocket(SSLSocketFactory.java:403)
	at org.apache.http.impl.conn.DefaultClientConnectionOperator.openConnection(DefaultClientConnectionOperator.java:177)
	at org.apache.http.impl.conn.ManagedClientConnectionImpl.open(ManagedClientConnectionImpl.java:304)
	at org.apache.http.impl.client.DefaultRequestDirector.tryConnect(DefaultRequestDirector.java:611)
	at org.apache.http.impl.client.DefaultRequestDirector.execute(DefaultRequestDirector.java:446)
	at org.apache.http.impl.client.AbstractHttpClient.doExecute(AbstractHttpClient.java:863)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:57)
	at org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine.invoke(ApacheHttpClient4Engine.java:283)
	... 38 more
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:387)
	at sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:292)
	at sun.security.validator.Validator.validate(Validator.java:260)
	at sun.security.ssl.X509TrustManagerImpl.validate(X509TrustManagerImpl.java:324)
	at sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:229)
	at sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:124)
	at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1491)
	... 55 more
Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
	at sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)
	at sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)
	at java.security.cert.CertPathBuilder.build(CertPathBuilder.java:280)
	at sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:382)
	... 61 more
[2016-02-28 16:58:11,426 INFO            o.o.p.r.n.NodeSource] [LocalNodes] Pinging alive nodes
```

A [quick search](https://www.google.fr/search?q=Caused+by%3A+sun.security.validator.ValidatorException%3A+PKIX+path+building+failed%3A+sun.security.provider.certpath.SunCertPathBuilderException%3A+unable+to+find+valid+certification+path+to+requested+target&ie=UTF-8) lets suppose it is a configuration issue on my side (e.g. certificate that is not trusted). However, the documentation gives no explanation about how to solve such a problem.

The default HTTPS configuration with provided keystore should work out-of-the box.

I performed the tests on Fedora 22 and OS X Yosemite with PWS 7.3.1 Enterprise version.

--- a/build.gradle
+++ b/build.gradle
@@ -11,6 +11,7 @@ subprojects {
     repositories {
         if (project.hasProperty('local')) mavenLocal()
         jcenter()
+        maven { url 'http://repository.activeeon.com/content/groups/proactive/' }
         maven { url 'http://repository.sonatype.org/content/groups/forge/' }
         maven { url 'http://repository.activeeon.com/content/repositories/thirdparty/' }
     }
@@ -44,12 +45,12 @@ subprojects {
         compile 'org.slf4j:slf4j-api:1.7.12'
         compile 'commons-io:commons-io:1.3.1'
         compile 'commons-fileupload:commons-fileupload:1.2'
-        compile 'org.apache.httpcomponents:httpclient:4.2.6'
+        compile 'org.apache.httpcomponents:httpclient:4.3.6'
         compile 'org.apache.commons:gwt-commons-patricia-trie:4.0'
-		
+        compile "org.ow2.proactive:common-http:${schedulingPortalVersion}"
 
         compile 'org.jboss.resteasy:resteasy-jaxrs:3.0.11.Final'
-        compile 'org.apache.httpcomponents:httpmime:4.2.6'
+        compile 'org.apache.httpcomponents:httpmime:4.3.6'
 
         runtime 'com.google.gwt:gwt-servlet:2.7.0'
 

--- a/build.gradle
+++ b/build.gradle
@@ -49,7 +49,7 @@ subprojects {
         compile 'org.apache.commons:gwt-commons-patricia-trie:4.0'
         compile "org.ow2.proactive:common-http:${schedulingPortalVersion}"
 
-        compile 'org.jboss.resteasy:resteasy-jaxrs:3.0.11.Final'
+        compile 'org.jboss.resteasy:resteasy-jaxrs:3.0.17.Final'
         compile 'org.apache.httpcomponents:httpmime:4.3.6'
 
         runtime 'com.google.gwt:gwt-servlet:2.7.0'
@@ -85,7 +85,7 @@ project(':applet') {
 project(':rm-portal') {
     dependencies {
         compile 'codemirror:codemirror2-gwt:1.2.0'
-        compile 'org.jboss.resteasy:resteasy-client:3.0.11.Final'
+        compile 'org.jboss.resteasy:resteasy-client:3.0.17.Final'
 
         runtime 'org.slf4j:slf4j-simple:1.7.12'
 
@@ -106,9 +106,8 @@ project(':scheduler-portal') {
         compile 'raphaelgwt:raphaelgwt:r28'
         compile 'org.vectomatic:lib-gwt-file:0.3.3'
         compile 'org.codehaus.jettison:jettison:1.2'
-        compile 'org.jboss.resteasy:resteasy-multipart-provider:3.0.11.Final'
+        compile 'org.jboss.resteasy:resteasy-multipart-provider:3.0.17.Final'
         compile 'org.apache.commons:commons-csv:1.1'
-
 
         compile project(':common-portal')
 

--- a/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/server/HttpUtils.java
+++ b/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/server/HttpUtils.java
@@ -36,47 +36,15 @@
  */
 package org.ow2.proactive_grid_cloud_portal.common.server;
 
-import org.apache.http.conn.scheme.Scheme;
-import org.apache.http.conn.ssl.SSLSocketFactory;
-import org.apache.http.conn.ssl.TrustStrategy;
-import org.apache.http.impl.client.DefaultHttpClient;
-import org.apache.http.impl.conn.PoolingClientConnectionManager;
-
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.security.cert.CertificateException;
-import java.security.cert.X509Certificate;
 
 
 public final class HttpUtils {
 
     private HttpUtils() {
-    }
-
-    public static DefaultHttpClient createDefaultExecutor() {
-        PoolingClientConnectionManager cm = new PoolingClientConnectionManager();
-        cm.setDefaultMaxPerRoute(50);
-        cm.setMaxTotal(50);
-        DefaultHttpClient httpClient = new DefaultHttpClient(cm);
-
-        try {
-            SSLSocketFactory socketFactory = new SSLSocketFactory(new RelaxedTrustStrategy(),
-              SSLSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
-            Scheme https = new Scheme("https", 443, socketFactory);
-            httpClient.getConnectionManager().getSchemeRegistry().register(https);
-        } catch (Exception ignored) {
-        }
-
-        return httpClient;
-    }
-
-    private static class RelaxedTrustStrategy implements TrustStrategy {
-        @Override
-        public boolean isTrusted(X509Certificate[] arg0, String arg1) throws CertificateException {
-            return true;
-        }
     }
 
     public static String convertToString(InputStream inputStream, boolean keepNewLines) throws IOException {
@@ -107,4 +75,5 @@ public final class HttpUtils {
     public static String convertToString(InputStream inputStream) throws IOException {
         return convertToString(inputStream, false);
     }
+
 }

--- a/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/server/MotdServlet.java
+++ b/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/server/MotdServlet.java
@@ -36,40 +36,40 @@
  */
 package org.ow2.proactive_grid_cloud_portal.common.server;
 
-import org.apache.commons.io.FileUtils;
-import org.apache.http.client.HttpClient;
-import org.apache.http.client.ResponseHandler;
-import org.apache.http.client.methods.HttpGet;
-import org.apache.http.impl.client.BasicResponseHandler;
-import org.apache.http.impl.client.DefaultHttpClient;
-import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import java.io.File;
+import java.io.IOException;
 
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import javax.ws.rs.core.Response;
 
-import java.io.File;
-import java.io.IOException;
+import org.ow2.proactive.http.HttpClientBuilder;
+import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
+import org.apache.commons.io.FileUtils;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.BasicResponseHandler;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * Return the content of the motd.txt file,
- * of the reponse of the *.motd.url if defined 
- * 
- * @author mschnoor
+ * of the reponse of the *.motd.url if defined
  *
+ * @author mschnoor
  */
 @SuppressWarnings("serial")
 public class MotdServlet extends HttpServlet {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(MotdServlet.class);
 
+    private static final String MOTD_FILE_NAME = "motd.txt";
+
     private static long lastModified = 0L;
     private static String fileContent = "";
-    private static final String motdFileName = "motd.txt";
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) {
@@ -77,13 +77,12 @@ public class MotdServlet extends HttpServlet {
         response.setContentType("text/html");
 
         try {
-
             String url = Config.get().getMotdUrl();
 
             // no MOTD URL : use local file
             if (url == null || url.trim().length() == 0) {
 
-                File f = new File(this.getServletContext().getRealPath(motdFileName));
+                File f = new File(this.getServletContext().getRealPath(MOTD_FILE_NAME));
                 long ft = f.lastModified();
                 if (ft != lastModified) {
                     lastModified = ft;
@@ -98,21 +97,19 @@ public class MotdServlet extends HttpServlet {
                 response.getWriter().write(fileContent);
 
             } else {
-                HttpClient httpclient = new DefaultHttpClient();
-
-                try {
+                try (CloseableHttpClient httpclient =
+                             new HttpClientBuilder()
+                                     .allowAnyCertificate(Config.get().isHttpsAllowAnyCertificate())
+                                     .allowAnyHostname(Config.get().isHttpsAllowAnyHostname()).build()) {
                     HttpGet httpget = new HttpGet(url);
 
                     ResponseHandler<String> responseHandler = new BasicResponseHandler();
                     String responseBody = httpclient.execute(httpget, responseHandler);
                     response.setStatus(Response.Status.OK.getStatusCode());
                     response.getWriter().write(responseBody);
-                }
-                catch (Exception e) {
+                } catch (Exception e) {
                     response.setStatus(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode());
                     response.getWriter().write("Server error");
-                } finally {
-                    httpclient.getConnectionManager().shutdown();
                 }
             }
 

--- a/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/shared/Config.java
+++ b/common-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/common/shared/Config.java
@@ -40,6 +40,7 @@ import java.util.Date;
 import java.util.HashMap;
 import java.util.Map;
 
+import org.ow2.proactive.web.WebProperties;
 import com.google.gwt.i18n.shared.DateTimeFormat;
 import com.google.gwt.i18n.shared.DefaultDateTimeFormatInfo;
 
@@ -61,6 +62,10 @@ public abstract class Config {
     private Map<String, String> backup = null;
 
     private static Config instance = null;
+
+    public static final String HTTPS_ALLOW_ANY_CERTIFICATE = WebProperties.WEB_HTTPS_ALLOW_ANY_CERTIFICATE;
+
+    public static final String HTTPS_ALLOW_ANY_HOSTNAME = WebProperties.WEB_HTTPS_ALLOW_ANY_HOSTNAME;
 
     protected Config() {
         this.properties = new HashMap<String, String>();
@@ -112,6 +117,24 @@ public abstract class Config {
      */
     public Map<String, String> getProperties() {
         return this.properties;
+    }
+
+    public boolean isHttpsAllowAnyCertificate() {
+        return getBooleanValue(HTTPS_ALLOW_ANY_CERTIFICATE, false);
+    }
+
+    public boolean isHttpsAllowAnyHostname() {
+        return getBooleanValue(HTTPS_ALLOW_ANY_HOSTNAME, false);
+    }
+
+    private boolean getBooleanValue(String property, boolean defaultValue) {
+        String value = this.properties.get(property);
+
+        if (value == null) {
+            return defaultValue;
+        }
+
+        return "true".equalsIgnoreCase(value);
     }
 
     /**

--- a/rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/server/RMServiceImpl.java
+++ b/rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/server/RMServiceImpl.java
@@ -36,33 +36,6 @@
  */
 package org.ow2.proactive_grid_cloud_portal.rm.server;
 
-import org.apache.commons.io.IOUtils;
-import org.apache.http.HttpResponse;
-import org.apache.http.client.methods.HttpPost;
-import org.apache.http.entity.mime.MultipartEntity;
-import org.apache.http.entity.mime.content.ByteArrayBody;
-import org.apache.http.entity.mime.content.FileBody;
-import org.apache.http.entity.mime.content.StringBody;
-import org.apache.http.impl.client.DefaultHttpClient;
-import org.jboss.resteasy.client.jaxrs.ResteasyClient;
-import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
-import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
-import org.ow2.proactive_grid_cloud_portal.common.server.ConfigReader;
-import org.ow2.proactive_grid_cloud_portal.common.server.ConfigUtils;
-import org.ow2.proactive_grid_cloud_portal.common.server.HttpUtils;
-import org.ow2.proactive_grid_cloud_portal.common.server.Service;
-import org.ow2.proactive_grid_cloud_portal.common.shared.RestServerException;
-import org.ow2.proactive_grid_cloud_portal.common.shared.ServiceException;
-import org.ow2.proactive_grid_cloud_portal.rm.client.RMService;
-import org.ow2.proactive_grid_cloud_portal.rm.shared.RMConfig;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import javax.management.MalformedObjectNameException;
-import javax.management.ObjectName;
-import javax.ws.rs.WebApplicationException;
-import javax.ws.rs.core.MediaType;
-import javax.ws.rs.core.Response;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
@@ -73,6 +46,35 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+
+import org.ow2.proactive.http.HttpClientBuilder;
+import org.ow2.proactive_grid_cloud_portal.common.server.ConfigReader;
+import org.ow2.proactive_grid_cloud_portal.common.server.ConfigUtils;
+import org.ow2.proactive_grid_cloud_portal.common.server.Service;
+import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
+import org.ow2.proactive_grid_cloud_portal.common.shared.RestServerException;
+import org.ow2.proactive_grid_cloud_portal.common.shared.ServiceException;
+import org.ow2.proactive_grid_cloud_portal.rm.client.RMService;
+import org.ow2.proactive_grid_cloud_portal.rm.shared.RMConfig;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.mime.MultipartEntity;
+import org.apache.http.entity.mime.content.ByteArrayBody;
+import org.apache.http.entity.mime.content.FileBody;
+import org.apache.http.entity.mime.content.StringBody;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.jboss.resteasy.client.jaxrs.ResteasyClient;
+import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
+import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
+import org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import static org.ow2.proactive_grid_cloud_portal.common.server.HttpUtils.convertToString;
 
@@ -95,13 +97,21 @@ public class RMServiceImpl extends Service implements RMService {
      */
     private ExecutorService threadPool;
 
-    private DefaultHttpClient httpClient;
+    private CloseableHttpClient httpClient;
 
     @Override
     public void init() {
         loadProperties();
 
-        httpClient = HttpUtils.createDefaultExecutor();
+        Config config = Config.get();
+
+        httpClient =
+                new HttpClientBuilder()
+                        .maxConnections(50)
+                        .allowAnyCertificate(config.isHttpsAllowAnyCertificate())
+                        .allowAnyHostname(config.isHttpsAllowAnyHostname())
+                        .useSystemProperties().build();
+
         threadPool = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
     }
 
@@ -117,7 +127,8 @@ public class RMServiceImpl extends Service implements RMService {
      * Loads properties defined in the configuration file and in JVM arguments.
      */
     private void loadProperties() {
-        RMConfig.get().load(ConfigReader.readPropertiesFromFile(getServletContext().getRealPath(RMConfig.CONFIG_PATH)));
+        RMConfig.get().load(
+                ConfigReader.readPropertiesFromFile(getServletContext().getRealPath(RMConfig.CONFIG_PATH)));
         ConfigUtils.loadSystemProperties(RMConfig.get());
     }
 
@@ -238,7 +249,7 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     private MultipartEntity createLoginPasswordSSHKeyMultipart(String login, String pass,
-                                                               String ssh) throws UnsupportedEncodingException {
+            String ssh) throws UnsupportedEncodingException {
         MultipartEntity entity = new MultipartEntity();
         entity.addPart("username", new StringBody(login));
         entity.addPart("password", new StringBody(pass));
@@ -253,9 +264,11 @@ public class RMServiceImpl extends Service implements RMService {
      * (non-Javadoc)
      * @see org.ow2.proactive_grid_cloud_portal.rm.client.RMService#createNodeSource(java.lang.String, java.lang.String, java.lang.String, java.lang.String[], java.lang.String[], java.lang.String, java.lang.String[], java.lang.String[])
      */
-    public String createNodeSource(final String sessionId, final String nodeSourceName, final String infrastructureType,
-                                   final String[] infrastructureParameters, final String[] infrastructureFileParameters, final String policyType,
-                                   final String[] policyParameters, final String[] policyFileParameters) throws RestServerException,
+    public String createNodeSource(final String sessionId, final String nodeSourceName,
+            final String infrastructureType,
+            final String[] infrastructureParameters, final String[] infrastructureFileParameters,
+            final String policyType,
+            final String[] policyParameters, final String[] policyFileParameters) throws RestServerException,
             ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
@@ -297,7 +310,8 @@ public class RMServiceImpl extends Service implements RMService {
      * (non-Javadoc)
      * @see org.ow2.proactive_grid_cloud_portal.rm.client.RMService#lockNodes(java.lang.String, java.util.Set)
      */
-    public String lockNodes(final String sessionId, Set<String> urls) throws RestServerException, ServiceException {
+    public String lockNodes(final String sessionId,
+            Set<String> urls) throws RestServerException, ServiceException {
         return executeFunction(new BiFunction<RestClient, Set<String>, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient, Set<String> strings) {
@@ -324,7 +338,8 @@ public class RMServiceImpl extends Service implements RMService {
      * (non-Javadoc)
      * @see org.ow2.proactive_grid_cloud_portal.rm.client.RMService#removeNode(java.lang.String, java.lang.String)
      */
-    public String removeNode(final String sessionId, final String url, final boolean force) throws RestServerException,
+    public String removeNode(final String sessionId, final String url,
+            final boolean force) throws RestServerException,
             ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
@@ -352,7 +367,8 @@ public class RMServiceImpl extends Service implements RMService {
      * (non-Javadoc)
      * @see org.ow2.proactive_grid_cloud_portal.rm.client.RMService#releaseNode(java.lang.String, java.lang.String)
      */
-    public String releaseNode(final String sessionId, final String url) throws RestServerException, ServiceException {
+    public String releaseNode(final String sessionId,
+            final String url) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -375,7 +391,8 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String getMBeanInfo(final String sessionId, String name, final List<String> attrs) throws RestServerException,
+    public String getMBeanInfo(final String sessionId, String name,
+            final List<String> attrs) throws RestServerException,
             ServiceException {
         try {
             final ObjectName obj = new ObjectName(name);
@@ -392,7 +409,8 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String getNodeMBeanInfo(final String sessionId, final String nodeJmxUrl, final String objectName, final List<String> attrs)
+    public String getNodeMBeanInfo(final String sessionId, final String nodeJmxUrl, final String objectName,
+            final List<String> attrs)
             throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
@@ -403,7 +421,9 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String getNodeMBeanHistory(final String sessionId, final String nodeJmxUrl, final String objectName, final List<String> attrs, final String timeRange) throws RestServerException, ServiceException {
+    public String getNodeMBeanHistory(final String sessionId, final String nodeJmxUrl,
+            final String objectName, final List<String> attrs,
+            final String timeRange) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -414,7 +434,7 @@ public class RMServiceImpl extends Service implements RMService {
 
     @Override
     public String getNodeMBeansInfo(final String sessionId, final String nodeJmxUrl, final String objectNames,
-                                    final List<String> attrs) throws RestServerException, ServiceException {
+            final List<String> attrs) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -424,8 +444,9 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String getNodeMBeansHistory(final String sessionId, final String nodeJmxUrl, final String objectNames,
-                                       final List<String> attrs, final String timeRange) throws RestServerException, ServiceException {
+    public String getNodeMBeansHistory(final String sessionId, final String nodeJmxUrl,
+            final String objectNames,
+            final List<String> attrs, final String timeRange) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -435,7 +456,8 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String getStatHistory(final String sessionId, final String range) throws RestServerException, ServiceException {
+    public String getStatHistory(final String sessionId,
+            final String range) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -445,7 +467,8 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     @Override
-    public String executeNodeScript(final String sessionId, final String script, final String engine, final String nodeUrl) throws RestServerException, ServiceException {
+    public String executeNodeScript(final String sessionId, final String script, final String engine,
+            final String nodeUrl) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
             public InputStream apply(RestClient restClient) {
@@ -455,14 +478,17 @@ public class RMServiceImpl extends Service implements RMService {
     }
 
     private RestClient getRestClientProxy() {
-        ResteasyClient client = new ResteasyClientBuilder().asyncExecutor(threadPool).build();
+        ResteasyClient client =
+                new ResteasyClientBuilder().asyncExecutor(threadPool)
+                        .httpEngine(new ApacheHttpClient4Engine(httpClient)).build();
+
         ResteasyWebTarget target = client.target(RMConfig.get().getRestUrl());
 
         return target.proxy(RestClient.class);
     }
 
     private String executeFunction(BiFunction<RestClient, Set<String>, InputStream> action, Set<String> urls,
-                                   String actionName) throws ServiceException, RestServerException {
+            String actionName) throws ServiceException, RestServerException {
 
         RestClient restClientProxy = getRestClientProxy();
 
@@ -483,13 +509,15 @@ public class RMServiceImpl extends Service implements RMService {
         }
 
         if (failures > 0) {
-            throw new RestServerException("Failed to " + actionName + " all requested nodes: " + success + " succeeded, " + failures + " failed.");
+            throw new RestServerException(
+                    "Failed to " + actionName + " all requested nodes: " + success + " succeeded, " + failures + " failed.");
         }
 
         return "";
     }
 
-    private String executeFunctionReturnStreamAsString(Function<RestClient, InputStream> function) throws ServiceException, RestServerException {
+    private String executeFunctionReturnStreamAsString(
+            Function<RestClient, InputStream> function) throws ServiceException, RestServerException {
         RestClient restClientProxy = getRestClientProxy();
 
         InputStream inputStream = null;

--- a/rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/shared/RMConfig.java
+++ b/rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/shared/RMConfig.java
@@ -41,10 +41,8 @@ import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
 
 /**
  * RM specific configuration
- * 
- * 
- * @author mschnoor
  *
+ * @author mschnoor
  */
 public class RMConfig extends Config {
 
@@ -109,7 +107,6 @@ public class RMConfig extends Config {
     /** prefix used to form the jmx url to monitor the node sources in the RM */
     public static final String RM_JMX_PREFIX = "rm.jmx.prefix";
     public static final String RM_JMX_PREFIX_DEFAULT = "service:jmx:rmi:///jndi/rmi://";
-    
     
     private static RMConfig instance = null;
 

--- a/rm-portal/src/main/webapp/rm.conf
+++ b/rm-portal/src/main/webapp/rm.conf
@@ -2,10 +2,10 @@ rm.rest.url=http://localhost:8080/rest
 
 # define whether hostname checking is performed or not when HTTPS
 # is used to communicate with the REST API
-web.https.allow_any_hostname=true
+#web.https.allow_any_hostname=true
 # define whether all kind of certificates (e,g. self-signed) are allowed
 # or not when HTTPS is used to communicate with the REST API
-web.https.allow_any_certificate=true
+#web.https.allow_any_certificate=true
 
 # Update rm.url to build a correct command line for node.jar
 rm.url=pnp://localhost:64738

--- a/rm-portal/src/main/webapp/rm.conf
+++ b/rm-portal/src/main/webapp/rm.conf
@@ -1,5 +1,12 @@
 rm.rest.url=http://localhost:8080/rest
 
+# define whether hostname checking is performed or not when HTTPS
+# is used to communicate with the REST API
+web.https.allow_any_hostname=true
+# define whether all kind of certificates (e,g. self-signed) are allowed
+# or not when HTTPS is used to communicate with the REST API
+web.https.allow_any_certificate=true
+
 # Update rm.url to build a correct command line for node.jar
 rm.url=pnp://localhost:64738
 # rm.client.refresh.time=3000
@@ -9,4 +16,3 @@ rm.url=pnp://localhost:64738
 rm.version=@portal_version@
 rm.monitoring.period=15000
 rm.monitoring.protocol=default
-

--- a/scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java
+++ b/scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/server/SchedulerServiceImpl.java
@@ -58,10 +58,11 @@ import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 import javax.ws.rs.core.Response.Status;
 
+import org.ow2.proactive.http.HttpClientBuilder;
 import org.ow2.proactive_grid_cloud_portal.common.server.ConfigReader;
 import org.ow2.proactive_grid_cloud_portal.common.server.ConfigUtils;
-import org.ow2.proactive_grid_cloud_portal.common.server.HttpUtils;
 import org.ow2.proactive_grid_cloud_portal.common.server.Service;
+import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
 import org.ow2.proactive_grid_cloud_portal.common.shared.RestServerException;
 import org.ow2.proactive_grid_cloud_portal.common.shared.ServiceException;
 import org.ow2.proactive_grid_cloud_portal.scheduler.client.JobUsage;
@@ -77,12 +78,13 @@ import org.apache.http.entity.mime.MultipartEntity;
 import org.apache.http.entity.mime.content.ByteArrayBody;
 import org.apache.http.entity.mime.content.FileBody;
 import org.apache.http.entity.mime.content.StringBody;
-import org.apache.http.impl.client.DefaultHttpClient;
+import org.apache.http.impl.client.CloseableHttpClient;
 import org.codehaus.jettison.json.JSONArray;
 import org.codehaus.jettison.json.JSONException;
 import org.jboss.resteasy.client.jaxrs.ResteasyClient;
 import org.jboss.resteasy.client.jaxrs.ResteasyClientBuilder;
 import org.jboss.resteasy.client.jaxrs.ResteasyWebTarget;
+import org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine;
 
 import static org.ow2.proactive_grid_cloud_portal.common.server.HttpUtils.convertToString;
 
@@ -95,7 +97,7 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
 
     private static final String ISO_8601_FORMAT = "yyyy-MM-dd'T'HH:mmZ";
 
-    private DefaultHttpClient httpClient;
+    private CloseableHttpClient httpClient;
 
     /**
      * Number of threads created for the threadPool shared by RestEasy client proxies.
@@ -111,7 +113,15 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
     public void init() {
         loadProperties();
 
-        httpClient = HttpUtils.createDefaultExecutor();
+        Config config = Config.get();
+
+        httpClient =
+                new HttpClientBuilder()
+                        .maxConnections(50)
+                        .allowAnyCertificate(config.isHttpsAllowAnyCertificate())
+                        .allowAnyHostname(config.isHttpsAllowAnyHostname())
+                        .useSystemProperties().build();
+
         threadPool = Executors.newFixedThreadPool(THREAD_POOL_SIZE);
     }
 
@@ -472,8 +482,8 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
-    
+
+
     /*
      * (non-Javadoc)
      *
@@ -505,8 +515,8 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
-    
+
+
     public String getTaskCentricByTag(final String sessionId, final String tag, final long fromDate, final long toDate,
                                       final boolean myTasks, final boolean pending, final boolean running,
                                       final boolean finished, final int offset, final int limit,
@@ -520,8 +530,8 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
-    
+
+
     @Override
     public String getJobTaskTagsPrefix(final String sessionId, final String jobId, final String prefix) throws RestServerException, ServiceException {
     	return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
@@ -531,7 +541,7 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
+
 
     /*
      * (non-Javadoc)
@@ -559,8 +569,8 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
-    
+
+
     public String getJobInfoDetails(final String sessionId, final String jobId) throws RestServerException, ServiceException {
         return executeFunctionReturnStreamAsString(new Function<RestClient, InputStream>() {
             @Override
@@ -569,7 +579,7 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
             }
         });
     }
-    
+
 
     /*
      * (non-Javadoc)
@@ -1153,7 +1163,9 @@ public class SchedulerServiceImpl extends Service implements SchedulerService {
     }
 
     private RestClient getRestClientProxy() {
-        ResteasyClient client = new ResteasyClientBuilder().asyncExecutor(threadPool).build();
+        ResteasyClient client =
+                new ResteasyClientBuilder().asyncExecutor(threadPool)
+                        .httpEngine(new ApacheHttpClient4Engine(httpClient)).build();
         ResteasyWebTarget target = client.target(SchedulerConfig.get().getRestUrl());
 
         return target.proxy(RestClient.class);

--- a/scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/shared/SchedulerConfig.java
+++ b/scheduler-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/shared/SchedulerConfig.java
@@ -42,10 +42,7 @@ import org.ow2.proactive_grid_cloud_portal.common.shared.Config;
 /**
  * Scheduler specific configuration
  *
- *
- *
  * @author mschnoor
- *
  */
 public class SchedulerConfig extends Config {
 

--- a/scheduler-portal/src/main/webapp/scheduler.conf
+++ b/scheduler-portal/src/main/webapp/scheduler.conf
@@ -1,7 +1,14 @@
-#must be accessible remotly but given the public address our hostname of the scheduler
+# must be accessible remotly but given the public address our hostname of the scheduler
 sched.rest.url=http://localhost:8080/rest
 
-#must be accessible remotly with the public address our hostname of the scheduler
+# define whether hostname checking is performed or not when HTTPS
+# is used to communicate with the REST API
+web.https.allow_any_hostname=true
+# define whether all kind of certificates (e,g. self-signed) are allowed
+# or not when HTTPS is used to communicate with the REST API
+web.https.allow_any_certificate=true
+
+# must be accessible remotely with the public address or hostname of the scheduler
 #sched.rest.public.url=http://localhost:8080/rest
 
 # sched.client.refresh.time=3000
@@ -13,10 +20,10 @@ sched.version=@portal_version@
 
 ### NOVNC CONFIGURATION ###
 
-# must be public, accessible from the client (http://host:port or https://host:port)
-# is is the location of the websocket proxy, we use in case of HTTPS to present the user the certificate warning
+# must be publicly accessible from the client (http://host:port or https://host:port)
+# the value corresponds to the location of the websocket proxy, it is used in case of HTTPS
 #sched.novnc.url=http\://localhost\:5900
 
-# must be public, accessible from the client, the URL where the client access point
+# must be publicly accessible from the client, the URL where the client accesses
 # it is the HTML page that runs the JS code of novnc, creating the web socket and displaying the screen output
 #sched.novnc.page.url=http\://localhost\:8080/rest/novnc.html

--- a/scheduler-portal/src/main/webapp/scheduler.conf
+++ b/scheduler-portal/src/main/webapp/scheduler.conf
@@ -3,10 +3,10 @@ sched.rest.url=http://localhost:8080/rest
 
 # define whether hostname checking is performed or not when HTTPS
 # is used to communicate with the REST API
-web.https.allow_any_hostname=true
+#web.https.allow_any_hostname=true
 # define whether all kind of certificates (e,g. self-signed) are allowed
 # or not when HTTPS is used to communicate with the REST API
-web.https.allow_any_certificate=true
+#web.https.allow_any_certificate=true
 
 # must be accessible remotely with the public address or hostname of the scheduler
 #sched.rest.public.url=http://localhost:8080/rest

